### PR TITLE
Initial commit - reputation improvements.

### PR DESCRIPTION
**Pull request by mark-terry**
Signed-off-by: mark-terry <mark.terry@consensys.net>

## PR description
Modifies the existing strike/point based behaviour to a more strict point based system.

## Fixed Issue(s)
Part of #3659 

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java [hunk 1]
```diff
@@ -200,12 +200,17 @@ public void removeChainEstimatedHeightListener(final long listenerId) {
...
   public void recordRequestTimeout(final int requestCode) {
     LOG.debug("Timed out while waiting for response from peer {}", this);
-    reputation.recordRequestTimeout(requestCode).ifPresent(this::disconnect);
+    reputation.recordRequestTimeout().ifPresent(this::disconnect);
   }
 
   public void recordUselessResponse(final String requestType) {
     LOG.debug("Received useless response for {} from peer {}", requestType, this);
-    reputation.recordUselessResponse(System.currentTimeMillis()).ifPresent(this::disconnect);
+    reputation.recordUselessResponse().ifPresent(this::disconnect);
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java:208]: it's a bit confusing to have the 2 methods recordUselessResponse and recordUselessAndDisconnect - can they be combined?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/PeerReputation.java [hunk 2]
```diff
@@ -31,62 +29,44 @@
...
       TimeUnit.MILLISECONDS.convert(1, TimeUnit.MINUTES);
 
   private final ConcurrentMap<Integer, AtomicInteger> timeoutCountByRequestType =
       new ConcurrentHashMap<>();
-  private final Queue<Long> uselessResponseTimes = new ConcurrentLinkedQueue<>();
 
-  private static final int DEFAULT_SCORE = 100;
-  private static final int SMALL_ADJUSTMENT = 1;
-  private static final int LARGE_ADJUSTMENT = 10;
+  private static final int DEFAULT_SCORE = 20;
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/PeerReputation.java:38]: would be interesting to see some benchmarks on how these numbers go on a testnet!

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/PeerReputation.java [hunk 3]
```diff
@@ -31,62 +29,44 @@
...
 
-  public Optional<DisconnectReason> recordRequestTimeout(final int requestCode) {
-    final int newTimeoutCount = getOrCreateTimeoutCount(requestCode).incrementAndGet();
-    if (newTimeoutCount >= TIMEOUT_THRESHOLD) {
-      LOG.debug("Disconnection triggered by repeated timeouts");
+  public Optional<DisconnectReason> recordRequestTimeout() {
+    score -= LARGE_ADJUSTMENT;
+    if (score <= 0) {
+      LOG.debug("Disconnection triggered by timeout");
       score -= LARGE_ADJUSTMENT;
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/PeerReputation.java:47]: does this mean we're decrementing the score twice here?

**Review by macfarla — COMMENTED** [review]
Added some comments. When can we run this up and see how it goes??

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java (hunk 1 above)
Comment (mark-terry) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java:208]: I'll clarify the method name a bit.

**Comment by mark-terry** [comment]
Closing due to revised approach.

**Pull request closed without merging**